### ISSUE TITLE
precision loss

### ISSUE THREAD

**Issue by nanmi** [Dao-AILab/flash-attention#507]
Great work, I implemented an attention module through `pytorch`, as follows, and then tested it on the A100 platform and found that the `test_flash_attn_output` function in `test_flash_attn.py` and my custom attention function will have an average precision loss of *0.001xxx* as a whole. In addition, `test_flash_attn_output` the accuracy of the attention module implemented using `GEMM` in the function of `FasterTransformer` has a large error. Have you tested these differences? In addition, do you have more flash attention details that need to be clarified?
```python
def class_attention_ref(q, k, v, is_attention_mask=False):
    if is_attention_mask:
        seq_len = q.shape[1]
        attention_mask = torch.tril(torch.ones((seq_len, seq_len), dtype=torch.bool)).view(1, 1, seq_len, seq_len)
        attention_mask = attention_mask.to(dtype=torch.float16)  # fp16 compatibility
        attention_mask = (1.0 - attention_mask) * torch.finfo(torch.float16).min

    head_dim = q.shape[-1]
    attn_weights = torch.matmul(q.transpose(1,2), k.transpose(1,2).transpose(2, 3)) / math.sqrt(head_dim)
    if is_attention_mask:
        attn_weights = attn_weights + attention_mask

    attn_weights = nn.functional.softmax(attn_weights, dim=-1, dtype=torch.float32)
    attn_output = torch.matmul(attn_weights, v.transpose(1,2))
    attn_output = attn_output.transpose(1, 2)
    return attn_output
```

**Comment by tridao** [comment]
You want to compare:
(Flash attention in fp16 - standard attention in fp32) vs (standard attention in fp16 - standard attention in fp32).

Usually FlashAttention gets *smaller* error (i.e. more accurate).

**Comment by nanmi** [comment]
Great, yes I've fixed this issue
